// ngx-show-hide-password/esm5/lib/show-hide-trigger.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, ErrorHandler } from '@angular/core';
import { ShowHideService } from './show-hide.service';
var ShowHideTriggerDirective = /** @class */ (function () {
    function ShowHideTriggerDirective(service, errorHandler) {
        this.service = service;
        this.errorHandler = errorHandler;
    }
    /**
     * @return {?}
     */
    ShowHideTriggerDirective.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (this.showHideTrigger) {
            this.service.toggleShow(this.showHideTrigger);
        }
        else {
            this.errorHandler.handleError(new Error("No input id found. Please read the docs!"));
        }
    };
    ShowHideTriggerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[showHideTrigger]'
                },] }
    ];
    /** @nocollapse */
    ShowHideTriggerDirective.ctorParameters = function () { return [
        { type: ShowHideService },
        { type: ErrorHandler }
    ]; };
    ShowHideTriggerDirective.propDecorators = {
        showHideTrigger: [{ type: Input }],
        onClick: [{ type: HostListener, args: ['click',] }]
    };
    return ShowHideTriggerDirective;
}());
export { ShowHideTriggerDirective };
if (false) {
    /** @type {?} */
    ShowHideTriggerDirective.prototype.showHideTrigger;
    /**
     * @type {?}
     * @private
     */
    ShowHideTriggerDirective.prototype.service;
    /**
     * @type {?}
     * @private
     */
    ShowHideTriggerDirective.prototype.errorHandler;
}
