// bn-ng-idle/bundles/bn-ng-idle.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('bn-ng-idle', ['exports', '@angular/core', 'rxjs'], factory) :
    (factory((global['bn-ng-idle'] = {}),global.ng.core,global.rxjs));
}(this, (function (exports,i0,rxjs) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var BnNgIdleService = (function () {
        function BnNgIdleService() {
            this.expired$ = new rxjs.Subject();
        }
        /**
         * @param {?} timeOutSeconds
         * @return {?}
         */
        BnNgIdleService.prototype.startWatching = /**
         * @param {?} timeOutSeconds
         * @return {?}
         */
            function (timeOutSeconds) {
                var _this = this;
                this.idle$ = rxjs.merge(rxjs.fromEvent(document, 'mosuemove'), rxjs.fromEvent(document, 'click'), rxjs.fromEvent(document, 'mousedown'), rxjs.fromEvent(document, 'keypress'), rxjs.fromEvent(document, 'DOMMouseScroll'), rxjs.fromEvent(document, 'mousewheel'), rxjs.fromEvent(document, 'touchmove'), rxjs.fromEvent(document, 'MSPointerMove'), rxjs.fromEvent(window, 'mousemove'), rxjs.fromEvent(window, 'resize'));
                this.timeOutMilliSeconds = timeOutSeconds * 1000;
                this.idleSubscription = this.idle$.subscribe(function (res) {
                    _this.resetTimer();
                });
                this.startTimer();
                return this.expired$;
            };
        /**
         * @return {?}
         */
        BnNgIdleService.prototype.startTimer = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.timer$ = rxjs.timer(this.timeOutMilliSeconds, this.timeOutMilliSeconds).subscribe(function (res) {
                    _this.expired$.next(true);
                });
            };
        /**
         * @return {?}
         */
        BnNgIdleService.prototype.resetTimer = /**
         * @return {?}
         */
            function () {
                this.timer$.unsubscribe();
                this.startTimer();
            };
        /**
         * @return {?}
         */
        BnNgIdleService.prototype.stopTimer = /**
         * @return {?}
         */
            function () {
                this.timer$.unsubscribe();
                this.idleSubscription.unsubscribe();
            };
        BnNgIdleService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] },
        ];
        /** @nocollapse */
        BnNgIdleService.ctorParameters = function () { return []; };
        /** @nocollapse */ BnNgIdleService.ngInjectableDef = i0.defineInjectable({ factory: function BnNgIdleService_Factory() { return new BnNgIdleService(); }, token: BnNgIdleService, providedIn: "root" });
        return BnNgIdleService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var BnNgIdleModule = (function () {
        function BnNgIdleModule() {
        }
        BnNgIdleModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [],
                        declarations: [],
                        providers: [BnNgIdleService],
                        exports: []
                    },] },
        ];
        return BnNgIdleModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.BnNgIdleService = BnNgIdleService;
    exports.BnNgIdleModule = BnNgIdleModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
