// ngx-show-hide-password/esm2015/lib/show-hide-password.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, Renderer2, ChangeDetectionStrategy } from '@angular/core';
import { faEye, faEyeSlash } from '@fortawesome/free-solid-svg-icons';
import { ShowHideService } from './show-hide.service';
/** @enum {string} */
const BtnStyle = {
    Primary: 'primary',
    Secondary: 'secondary',
    Success: 'success',
    Danger: 'danger',
    Warning: 'warning',
    Info: 'info',
    Dark: 'dark',
    Light: 'light',
};
export { BtnStyle };
/**
 * Add a split input button to password or text input. Toggles input type between "text" and "password".
 *
 * \@example
 * <show-hide-password size="sm|lg">
 * <input type="password" name=... />
 * </show-hide-password>
 */
export class ShowHidePasswordComponent {
    /**
     * @param {?} service
     * @param {?} elem
     * @param {?} renderer
     */
    constructor(service, elem, renderer) {
        this.service = service;
        this.elem = elem;
        this.renderer = renderer;
        this.btnStyle = BtnStyle.Secondary;
        this.btnOutline = true;
        this.faEye = faEye;
        this.faEyeSlash = faEyeSlash;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.input = this.elem.nativeElement.querySelector('input');
        if (!this.input) {
            throw new Error(`No input element found. Please read the docs!`);
        }
        this.id = this.input.getAttribute('id');
        if (!this.id) {
            this.id = 'showHideInput' + Math.round(Math.random() * 100000);
            this.renderer.setAttribute(this.input, 'id', this.id);
        }
        this.renderer.addClass(this.elem.nativeElement, 'input-group');
        if (this.size === 'sm') {
            this.renderer.addClass(this.elem.nativeElement, 'input-group-sm');
        }
        else if (this.size === 'lg') {
            this.renderer.addClass(this.elem.nativeElement, 'input-group-lg');
        }
        this.isHidden = this.input.type === 'password';
        this.renderer.addClass(this.input, 'form-control'); // just to be sure
        this.service.setShow(this.id, this.input.type !== 'password');
        this.subscription = this.service.getObservable(this.id).subscribe((/**
         * @param {?} show
         * @return {?}
         */
        show => {
            this.isHidden = !show;
            this.renderer.setAttribute(this.input, 'type', show ? 'text' : 'password');
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
ShowHidePasswordComponent.decorators = [
    { type: Component, args: [{
                selector: 'show-hide-password',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
  <ng-content></ng-content>
  <div class="input-group-append ngx-show-hide-password">
    <button class="btn" [ngClass]="(btnOutline ? 'btn-outline-' + btnStyle : 'btn-' + btnStyle)"
      type="button" [showHideTrigger]="id">
      <fa-icon [fixedWidth]="true" size="lg" [icon]="(isHidden ? faEye : faEyeSlash)"
        [showHideStatus]="{id: id}"></fa-icon>
    </button>
  </div>
`
            }] }
];
/** @nocollapse */
ShowHidePasswordComponent.ctorParameters = () => [
    { type: ShowHideService },
    { type: ElementRef },
    { type: Renderer2 }
];
ShowHidePasswordComponent.propDecorators = {
    btnStyle: [{ type: Input }],
    btnOutline: [{ type: Input }],
    size: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ShowHidePasswordComponent.prototype.subscription;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.btnStyle;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.btnOutline;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.size;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.input;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.isHidden;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.id;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.faEye;
    /** @type {?} */
    ShowHidePasswordComponent.prototype.faEyeSlash;
    /**
     * @type {?}
     * @private
     */
    ShowHidePasswordComponent.prototype.service;
    /**
     * @type {?}
     * @private
     */
    ShowHidePasswordComponent.prototype.elem;
    /**
     * @type {?}
     * @private
     */
    ShowHidePasswordComponent.prototype.renderer;
}
